### ISSUE TITLE
[HW7] Stack Pointer Allocation

### ISSUE THREAD

**Issue by Ngiong** [kaist-cp/cs420#211]
Hello,

1. Can you explain how can I determine how much I should allocate stack?
I thought we are storing both return address (ra) and frame pointer (s0) in the new allocated stack (after decrement the stack pointer)?
So is it 8 (ra) + 8 (s0) + total_size(arguments) + total_size(local allocations)?

2. There are 2 ways to call a function: (1) allocating stack & (2) passing through a0, a1, a2, ... (if argument <= 8). And we know that if the function has Call Instruction, we must use the (1). 
To satisfy hw7 (generate correct RISC-V code), **is it possible to somehow "generalize" to only use only the (1)**?

3. _This might be a dumb question_. If our KECC IR's block consists of N instructions (which is N temporary registers), can we "increase how much we allocate the stack" such that every temporary register **can be stored there**, and **can be accessed/read anytime**? Will it work?
I'm thinking that if we only use 2-3 CPU registers for two operands & the result, we might lose some information if our IR happens to access early temporary register (?)

4. Do we have to really understand RISC-V syntaxes if we want to complete this homework?

Please, correct me if I'm wrong.
I'm really new at this machine-level code generation, so maybe my question might have misunderstandings.

Thank you.

**Comment by jeehoonkang** [comment]
Regarding point 2., let's continue the discussion at #209.

**Comment by cmpark0126** [comment]
Thank you for the question.

1. Allocate as many stacks as necessary to execute the function. For example, if the register is insufficient to execute the function, some values need to be backed up to the stack sometimes. In this case, you need to allocate some memory to the stack for this value.

    FYI, I suggest the following formula: 8(ra) + 8(s0) + total_size(parameters passed through memory) + total_size(values needed to be backed up the middle of function execution) + total_size(local allocations). Do not violate the calling convention. It will be discussed at #209.

2. Please refer to #209

3. I believe this question can be resolved at answer1 above. If you have further questions, please ask again.

4. You don't have to understand everything. FYI, it is recommended to be familiar with the RISC-V instructions & directives given for debugging.

**Comment by Ngiong** [comment]
Thank you for the answer @cmpark0126

**Comment by Ngiong** [comment]
By the way, will there any code walkthrough video to explain the syntax/semantics of `asm/mod.rs`?
So now I'm pretty don't know what to do to start HW7, and I planned to "manually" construct asm::TranslationUnit from godbolt's output for `fibonacci.c` (I hope this is a good example for starter (?))

But I'm confused about what is Section, and why there are 2 Vec<Directive> when I tried to construct a variable. And I noticed that Immediate::Value can only receive unsigned value (although from godbolt, it seems negative values are permitted)?

```
pub enum Immediate {
    // TODO: consider architecture dependency (current: 64-bit architecture)
    Value(u64),
```
